### PR TITLE
feat(cli): batch rendering — one output per variables row with manifest

### PR DESCRIPTION
**Pull request by mvanhorn**
## What

Data-driven batch rendering: one composition plus a JSON array of variable rows produces one video per row.

```bash
hyperframes render --batch rows.json --output "renders/{name}.mp4"
```

Output names are templated from row keys (plus `{index}`), every row is validated up front (strict-variables aware, collision-checked) before the first render starts, and a `manifest.json` records per-row status, output path, variables, and timing. `--batch-concurrency N` (default 1) and `--batch-fail-fast` control execution; `--json` emits machine-readable progress events.

Three rows in, three personalized videos out (one invocation, rendered by this branch):

batch render demo

## Why

The variables system already parametrizes a single render (`--variables`, `--variables-file`, threaded through plan() and renderChunk() in #962 and the render server in #1152), and the README pitches "reusable motion graphics for automated content pipelines" — but the pipeline half is missing: rendering the same composition once per data row means a shell loop, N cold CLI startups, no pre-flight validation, and no machine-readable result. Dataset rendering is a documented core Remotion workflow; agents producing localized variants, personalized outros, or one-clip-per-product catalogs hit this shape constantly.

## How

- `packages/cli/src/commands/batchRender.ts` (new, ~460 lines): row parsing (`[...]` or `{"rows": [...]}`), output templating with placeholder validation, collision detection (duplicate paths and manifest clashes fail before any render), per-row variable validation reusing `validateVariablesAgainstProject`, a bounded-concurrency runner, and the manifest writer. Rows that fail record their error and the batch continues unless `--batch-fail-fast`.
- `packages/cli/src/commands/render.ts`: registers the flags, rejects `--batch` combined with `--variables`/`--variables-file`, and lazy-imports the batch module only when `--batch` is present — the single-render path is untouched.
- Default `--batch-concurrency 1` is deliberate: a single render already saturates the worker pool, so rows run sequentially unless the compositions are small.
- Docs: batch section in `docs/guides/rendering.mdx`, variables concept page, pipeline guide cross-link.

## Test plan

The demo above is real output: a 3-row batch (`q3-report`, `q4-report`, `year-recap`) rendered with this branch in one invocation, each video picking up its row's title and accent color, with `manifest.json` reporting `total: 3, completed: 3, failed: 0`.

- [x] Unit tests added/updated (row parsing shapes, template resolution and missing-key errors, collision detection, strict-variables per-row failures, manifest lifecycle, fail-fast vs continue-on-error)
- [x] Manual testing performed (end-to-end 3-row batch above)
- [x] Documentation updated (rendering guide, variables concept, pipeline guide)

### CODE REVIEW

**Comment by jrusso1020** [comment]
Ohhhh very interesting idea @mvanhorn , let me take a deeper look to understand this and review in detail! 

Thank you once again for the contribution

**Comment by mvanhorn** [comment]
Thanks! The quickest way to see the shape is the demo's manifest.json - 3 rows in, 3 templated outputs with per-row status and timing. The review weight is almost all in batchRender.ts (preflight validation + the runner); render.ts only gains flag parsing and a lazy import, so the single-render path is untouched.

**Comment by jrusso1020** [comment]
@mvanhorn okay took a deeper look at the code myself and think I'm over aligned with this! Thanks for the contribution definitely seems helpful for batch rendering workflows locally with variable inputs

let me have my agent give a second pass at the code quality and I'll think about the interface design as well

File: packages/cli/src/commands/batchRender.ts [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+}
+
+export interface PreparedBatchRender {
+  batchPath: string;
+  manifestPath: string;
+  variableIssueCount: number;
+  rows: PreparedBatchRow[];
+}
+
+export interface BatchRenderResult {
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:36]: **naming/dedup (also the `fallow/unused-type` finding):** `BatchRenderResult` is structurally identical to `SingleRenderResult` (`render.ts:807`), and the name reads like the result of the *whole batch* — but that's `BatchManifest`; this is the per-row result. Suggest one shared type: either import `SingleRenderResult` here, or rename this to `RowRenderResult` and have `render.ts` import it. Either direction also clears the Fallow `unused-type` major.

File: packages/cli/src/commands/batchRender.ts [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  durationMs?: number;
+  renderTimeMs: number;
+}
+
+export interface BatchManifestRow {
+  index: number;
+  outputPath: string;
+  status: "pending" | "running" | "completed" | "failed" | "skipped";
+  durationMs: number | null;
+  renderTimeMs: number | null;
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:46]: **important — settle the field names before `version: 1` ships:** in the manifest, `durationMs` (composition length) vs `renderTimeMs` (wall clock) is genuinely ambiguous — a consumer can't tell which one is "how long did this take" without reading this source. Since the manifest is an explicitly versioned public contract, renaming later is a breaking change; now is the cheap moment. Suggest `compositionDurationMs` (or `videoDurationMs`) alongside `renderTimeMs` — or `wallTimeMs` if you want the pair fully self-describing. Same applies to `BatchRenderResult`/`SingleRenderResult`.

File: packages/cli/src/commands/batchRender.ts [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+  if (!Object.hasOwn(row, key)) {
+    throw new BatchRenderInputError(
+      "Invalid output template",
+      `Missing value for placeholder {${key}} in row ${index}.`,
+    );
+  }
+
+  const value = row[key];
+  if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
+    return String(value);
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:136]: **important — placeholder values flow into paths unsanitized.** A row like `{"name": "../../somewhere/else"}` or a value containing `/` silently writes outside (or nests under) the intended output tree — `mkdirSync(..., {recursive: true})` will happily create whatever it resolves to. Fine when rows are hand-written, but this feature explicitly targets agent/pipeline data sources where row values may be third-party (product names, user names). Suggest rejecting path separators and `..` segments in placeholder *values* by default (template authors who want subdirs can put the `/` in the template itself, e.g. `renders/{lang}/{name}.mp4`).

File: packages/cli/src/commands/batchRender.ts [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+
+  const variableRows = parseBatchRows(raw, batchPath);
+  const rows = variableRows.map((variables, index) => ({
+    index,
+    variables,
+    outputPath: resolve(resolveOutputTemplate(options.outputTemplate, variables, index)),
+  }));
+  const manifestPath = join(
+    commonOutputDirectory(rows.map((row) => row.outputPath)),
+    "manifest.json",
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:254]: **important — can clobber a pre-existing, unrelated `manifest.json`.** The manifest lands at the common output directory under a fixed name, and `checkOutputCollisions` only guards batch outputs against it — not files already on disk. `--output "public/{name}.mp4"` in a web project would silently overwrite a PWA manifest. Two cheap mitigations, either or both: (1) if `manifest.json` exists and doesn't parse as a batch manifest (`version` field), fail preflight; (2) add a `--batch-manifest <path>` override — which also solves two concurrent batches sharing an output dir.

File: packages/cli/src/commands/batchRender.ts [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+  return template.replace(PLACEHOLDER_RE, (_match, key: string) =>
+    placeholderValue(row, key, index),
+  );
+}
+
+function isSameOrChildPath(path: string, parent: string): boolean {
+  return path === parent || path.startsWith(parent.endsWith(sep) ? parent : parent + sep);
+}
+
+export function commonOutputDirectory(outputPaths: readonly string[]): string {
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:159]: Fallow `unused-export` (major): nothing imports `commonOutputDirectory` — the tests exercise it indirectly through `prepareBatchRender`'s `manifestPath` assertions. Drop the `export` keyword to clear the finding.

File: packages/cli/src/commands/batchRender.ts [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+          )
+        : c.success(`Batch complete: ${manifest.completed} completed.`),
+    );
+    console.log(c.dim(`Manifest: ${manifest.manifestPath}`));
+  }
+
+  return manifest;
+}
+
+export function exitBatchRenderInputError(error: unknown): never {
```
Comment (jrusso1020) [inline packages/cli/src/commands/batchRender.ts:458]: Fallow `unused-export` (major) fires here too, but it's a **false positive** — this is called via the lazy `await import("./batchRender.js")` at `render.ts:587`, which Fallow's static import graph can't trace. Needs a suppression annotation (the repo's `fallow-ignore-next-line` convention) rather than a code change, unless you'd rather make the import static at the top of `render.ts` and give up the lazy-load.

File: packages/cli/src/commands/render.ts [hunk 7]
```diff
@@ -248,6 +252,26 @@ export default defineCommand({
...
+      type: "string",
+      description:
+        "Maximum number of batch rows to render at once. Default: 1, because each render already parallelizes across workers.",
+    },
+    "batch-fail-fast": {
+      type: "boolean",
+      description: "Stop launching new batch rows after the first row failure.",
+      default: false,
+    },
+    json: {
```
Comment (jrusso1020) [inline packages/cli/src/commands/render.ts:270]: **interface consistency:** `--json` without `--batch` is accepted and silently does nothing, while its siblings `--batch-concurrency`/`--batch-fail-fast` hard-error without `--batch` (lines 478-486). Two coherent options: (a) guard it identically, or (b) keep the generic name as a deliberate reservation for future machine-readable single-render output — but then the silent no-op should probably at least warn today. If batch-only is the intent, `--batch-json` would make the scoping self-evident. (a) is the smallest change.

**Review by jrusso1020 — APPROVED** [review]
Second pass focused on code quality + interface/naming, as @jrusso1020 mentioned above. I read all six files end-to-end (not just the diff) and verified the test-plan claims against the source: the preflight ordering (rows parsed and validated before any browser work, `render.ts:573-589`), the manifest lifecycle, and the fail-fast/skip semantics all match the description. The variables claim checks out too — and batch actually improves on the wrapper by loading the schema once per batch instead of once per row (`batchRender.ts:197`).

**Strengths worth calling out**
- The preflight-everything-first design (`prepareBatchRender`) is the right contract: collisions, bad templates, and strict-variable failures cost milliseconds instead of minutes of wasted renders.
- The runner (`batchRender.ts:404-417`) is a clean bounded-concurrency pool — no deps, no locks, correct under JS single-threading.
- `throwOnError`/`skipFeedback`/`exitAfterComplete` opt-ins (`render.ts:847-850`) are a surgical way to make the single-render path drivable as a library without touching its CLI behavior.
- Manifest journaling (rewrite after every state change, `version: 1` field) is the right shape for agent pipelines.

## Relaying the Fallow audit results

The Fallow CI job failed, but its sticky comment couldn't post (`Resource not accessible by integration` — fork PRs get a read-only token), so you can't see the findings. I reproduced them locally on this branch:

| Severity | Rule | Location | Finding |
|---|---|---|---|
| major | `fallow/unused-type` | `batchRender.ts:36` | `BatchRenderResult` never imported elsewhere |
| major | `fallow/unused-export` | `batchRender.ts:159` | `commonOutputDirectory` never imported elsewhere |
| major | `fallow/unused-export` | `batchRender.ts:458` | `exitBatchRenderInputError` never imported elsewhere |
| minor | `fallow/high-crap-score` | `batchRender.ts:196` | `validateBatchVariables` CRAP 31.6 (threshold 30, cyclomatic 10) |

The third one is a **false positive** — it IS used, via the lazy `await import("./batchRender.js")` at `render.ts:587`, which Fallow's static import graph can't see. Details inline. Fallow isn't a required check, but it'll stay red until these are suppressed/fixed.

## Interface & naming (the design pass)

- **Flag family**: `--batch <path>` taking a file sits slightly against the `--variables-file <path>` precedent. I wouldn't rename it — `--batch rows.json` reads well and the `--batch-*` prefix family (`--batch-concurrency`, `--batch-fail-fast`) is coherent — but worth deciding deliberately since it sets the pattern for future file-taking flags.
- **`--json` scoping** is the one real inconsistency in the flag surface (inline at `render.ts:270`).
- **Template syntax** `{key}` + built-in `{index}` matches ecosystem conventions and is the right choice. Two unobservable corners worth a docs line each: there's no escape for a literal `{...}` in an output path, and `{a.b}` looks like nested access but is a literal key lookup (`PLACEHOLDER_RE` allows dots).
- **Manifest field naming** is the one naming issue I'd actually settle before merge, because `manifest.json` is a versioned public contract — renaming fields later is a breaking change (inline at `batchRender.ts:46`).
- **Default `--batch-concurrency 1`** with the saturation rationale documented in the flag description and docs: good default, well justified.

## Nits (no action required)

- `batchRender.ts:383` — skipped rows put an explanation string in `error`. Consumers doing `rows.filter(r => r.error)` will count skips as failures; `status` already distinguishes. Consider `error: null` for skips or a separate field.
- `batchRender.ts:304-306` — `writeFileSync` isn't atomic; a crash mid-write can leave truncated JSON, which slightly undercuts the journal guarantee. Temp-file + rename is a two-line fix if you want it bulletproof.
- Docker mode runs one container per row, so the "N cold starts" cost partially returns there — fine, maybe worth a sentence in the rendering guide.
- `handleRenderError` with `throwOnError` throws `new Error(message)` and drops the `hint` — batch manifest errors lose the actionable hint text single-render users see.

## Explainer video

As part of this review I put together a HyperFrames explainer of the feature (built with the published 0.6.90 CLI — lint/validate/inspect clean): the before/after, the one-command flow, the preflight gates, and the run-loop journaling the manifest.

---

**Verdict:** APPROVE
**Reasoning:** All required checks are green and I found no correctness blockers — the importants are contract-naming and input-hygiene decisions best settled pre-merge, plus mechanical Fallow fixes (the failing job is non-required and couldn't post its own findings here).

— Claude Code (review pass for @jrusso1020)

**Review by miguel-heygen — CHANGES_REQUESTED** [review]
**Strengths:**
- `packages/cli/src/commands/batchRender.ts`: the preflight-before-first-render design is exactly right — collision detection, strict-variable validation, and output-path templating all fail fast before any video starts. This matches the CLI's existing "validate first, render second" pattern.
- `batchRender.test.ts`: the test suite covers the most important edge cases cleanly: collision detection, fail-fast skipping, continue-on-failure, and JSON event sequence. `prepareBatchRender` is cleanly unit-testable because `renderOne` is injected.

**Blocker — merge conflict:**
The PR is currently in `DIRTY` state (merge conflict against main). Needs a rebase before merge.

**Fallow audit note:**
The `Fallow audit` CI job shows as failed but that's a false alarm — the sticky-comment step got `Resource not accessible by integration` (a GitHub token permissions issue on the workflow, not a fallow finding). The actual fallow output is exclusively `WARN unknown fallow suppression issue kind` lines for pre-existing unrecognized suppression keys. No real findings.

**Nit — manifest written before row completes:**
`packages/cli/src/commands/batchRender.ts`: the manifest is persisted after every row (good for crash safety), but if `--batch-concurrency > 1`, rows complete out of order. The manifest rows array should be indexed by `row.index` when writing, not the completion order. Looking at the code this appears to be handled correctly (rows are pre-populated by index and updated in place), but worth confirming in a quick read.

**Nit — `--batch` combined with `--variables` is rejected at runtime:**
`render.ts` rejects `--batch` + `--variables` with a clear error, which is correct. Consider documenting this constraint in the `--batch` help text too so it's discoverable before invocation.

One existing approval from @jrusso1020. No blockers once rebased.

**Verdict:** REQUEST CHANGES (rebase only — no code issues)
**Reasoning:** The implementation is solid; the only gate is the merge conflict. Once rebased this is ready.

— magi

**Comment by mvanhorn** [comment]
@miguel-heygen rebased onto `main` and force-pushed (`93074c8`). The conflict was in `renderLocal` (`render.ts`): `main` had moved the FFmpeg check into the new `runEnvironmentChecks` preflight, while this branch changed the return type to `SingleRenderResult` so `batchRender` can consume per-row results. Resolved by keeping main's preflight body and this branch's return type; the post-preflight `loadProducer()` is now a single declaration.

Verified locally on the rebased branch: full CLI suite 767/767 (incl. `render.test.ts` + the 13 `batchRender` tests), `tsc --noEmit` clean, oxlint + oxfmt clean. PR is now `MERGEABLE`.

On @jrusso1020's pre-merge notes: happy to fold the two contract items in now if you'd like them before `version: 1` ships - the `durationMs`/`renderTimeMs` manifest field rename and sanitizing placeholder values that flow into output paths (e.g. a row value containing `/` or `..`). Otherwise I'll keep this PR rebase-only as you suggested and open a fast-follow. Your call.

**Review by miguel-heygen — APPROVED** [review]

**Comment by miguel-heygen** [comment]
### Merge activity

* **Jun 13, 5:56 AM UTC**: [Graphite](https://app.graphite.com/github/pr/heygen-com/hyperframes/1336) couldn't merge this PR because it failed for an unknown reason (Fast-forward merges are not supported for forked repositories. Please create a branch in the target repository in order to merge).

**Comment by mvanhorn** [comment]
Really appreciate the review and merge, @miguel-heygen. Data-driven batch rendering with one video per variables row was a fun one to build.
